afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+pub struct Socket {
+    fd: libc::c_int,
+}
+
+/// Represents a bound `AF_PACKET` socket. At this phase of a socket's lifecycle, it can be read
+/// to/written from.
+pub struct BoundSocket {
+    fd: libc::c_int,
+    send_addr: libc::sockaddr_ll,
+}
+
+impl Socket {
+    /// Creates a new unbound socket.
+    pub fn new() -> io::Result<Self> {
+        // This block is unsafe because it invokes a syscall. It is safe because it checks the
+        // syscall return value and returns an Err if the sycall fails. It does not interact with
+        // Rust-owned memory, so it's safe as per the borrow checker's rules.
+        let fd = unsafe {
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#socket
+            // man 7 packet
+            let fd = libc::socket(libc::AF_PACKET, libc::SOCK_RAW, libc::ETH_P_ALL.to_be());
+            if fd < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            fd
+        };
+        Ok(Self { fd })
+    }
+
+    /// Binds the socket to a network interface. This function consumes the `Socket` instance, as
+    /// no more configuration options may be safely changed.
+    pub fn bind(self, iface: impl AsRef<CStr>) -> io::Result<BoundSocket> {
+        // This block is unsafe because it invokes syscalls. It is safe because the syscall return
+        // values are checked at each syscall and returned as an Err value. This block should be
+        // tested with miri to confirm that the use of the &CStr it takes as input is correct.
+        let send_addr = unsafe {
+            // get the index of the interface
+            let mut ifr : linux::ifreq = MaybeUninit::zeroed().assume_init();
+            ptr::copy_nonoverlapping(iface.as_ref().as_ptr(),
+                ifr.ifr_ifrn.ifrn_name.as_mut_ptr(), libc::IFNAMSIZ);
+            // ioctl(SIOCGIFINDEX) fills in the index field of the ifreq object
+            // Resources:
+            // man 7 netdevice
+            let err = libc::ioctl(self.fd, linux::SIOCGIFINDEX, &ifr);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+
+            // bind the socket
+            let mut ll : libc::sockaddr_ll = MaybeUninit::zeroed().assume_init();
+            ll.sll_family = libc::AF_PACKET as libc::c_ushort;
+            ll.sll_ifindex = ifr.ifr_ifru.ifru_ivalue; // expanded from `ifr_ifindex` in kernel headers
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#bind
+            // man 7 packet regarding sockaddr_ll
+            let err = libc::bind(self.fd,
+                                 &mut ll as *mut _ as *mut libc::sockaddr,
+                                 mem::size_of::<libc::sockaddr_ll>() as libc::c_uint);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            ll
+        };
+        let fd = self.fd;
+        mem::forget(self);
```

Comment:
Why not `impl Drop` for `Socket`?

Reply:
The idea is that this function *consumes* the `Socket`, but hands its `fd` to a `BoundSocket`. So now the `BoundSocket` is responsible for closing the `fd`.

Without calling `mem::forget(self)`, `self` (the `Socket`) has its destructor run at the end of `bind`, which would close the `fd` before client code uses it in `BoundSocket`.

Reply:
Could we also represent the state machine transition by having a `send_addr: Option<libc::sockaddr_ll>,` in `Socket`? We wouldn't need to deal with this memory management stuff, but having `Socket` and `BoundSocket` distinct really makes it hard to use the API incorrectly (e.g. you can't `send` on a socket that isn't bound.). I'm not sure what the right design here is.

Reply:
That's what my original design did, but it made it possible to call `send` on an unbound socket, which is not ideal.

Reply:
Oh I see. There's no way to drop the consumed `Socket` here. The `mem::forget` documentation is a little scary:

> Any resources the value manages, such as heap memory or a file handle, will linger forever in an unreachable state.

It would be nice to know that the memory is actually being released properly without closing the file descriptor...

Reply:
The other options I can think of are

1. Add a `close` method and require the client to explicitly call it to close the file descriptor.
2. Use bind to a different fd for `BoundSocket` and drop the first one.
3. Convince me that the `mem::forget` option isn't as bad as I think.

Reply:
The `mem::forget` option is safe, because the `Socket` object manages the file descriptor. That file descriptor was moved to `BoundSocket`, which enables it to handle the lifecycle of the file descriptor.